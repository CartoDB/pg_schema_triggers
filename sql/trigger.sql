CREATE EXTENSION schema_triggers;

-- Create an event trigger for the trigger_create event.
CREATE FUNCTION on_trigger_create()
 RETURNS event_trigger
 LANGUAGE plpgsql
 AS $$
	DECLARE
		event_info SCHEMA_TRIGGERS.TRIGGER_CREATE_EVENTINFO;
	BEGIN
		event_info := schema_triggers.get_trigger_create_eventinfo();
		RAISE NOTICE 'on_trigger_create: (tgname=''%'', tgrelid=''%'', tgenabled=''%'', is_internal=''%'')',
			(event_info.new).tgname, (event_info.new).tgrelid::REGCLASS,
			(event_info.new).tgenabled, event_info.is_internal;
		IF (event_info.new).tgname LIKE 'test_%' THEN
			RAISE EXCEPTION 'trigger name cannot begin with "test_"';
		END IF;
		IF (event_info.new).tgrelid = 'untriggered'::REGCLASS THEN
			RAISE EXCEPTION 'cannot add triggers to relation named "untriggered".';
		END IF;
	END;
$$;
CREATE EVENT TRIGGER trigcreate ON trigger_create
	EXECUTE PROCEDURE on_trigger_create();

-- Create an event trigger for the trigger_adjust event.
-- TODO

-- Create an event trigger for the trigger_rename event.
-- TODO

-- Create an event trigger for the trigger_drop event.
-- TODO

-- Create some empty tables.
CREATE TABLE foo();
CREATE TABLE foobar();
CREATE TABLE untriggered();

-- Create a do-nothing trigger function for the following tests.
CREATE FUNCTION ignore()
 RETURNS trigger
 LANGUAGE plpgsql
 AS $$ BEGIN RETURN NULL; END; $$;

-- Create some triggers on the tables.
CREATE TRIGGER foo_trig BEFORE INSERT ON foo
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER foobar_trig BEFORE UPDATE OR DELETE ON foobar
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER test_will_not_work BEFORE INSERT ON foobar
	EXECUTE PROCEDURE ignore();
CREATE TRIGGER also_will_not_work BEFORE INSERT ON untriggered
	EXECUTE PROCEDURE ignore();

-- Rename, enable, and disable the triggers.
ALTER TABLE foo DISABLE TRIGGER foo_trig;
ALTER TRIGGER foo_trig ON foo RENAME TO foo_trig2;
ALTER TABLE foo ENABLE TRIGGER foo_trig2;
ALTER TABLE foobar ENABLE REPLICA TRIGGER foobar_trig;
ALTER TABLE foobar ENABLE ALWAYS TRIGGER foobar_trig;
ALTER TABLE foobar DISABLE TRIGGER foobar_trig;

-- Drop the triggers.
DROP TRIGGER foo_trig2 ON foo;
DROP TRIGGER foobar_trig ON foobar;

-- Create a table with an automatically-created constraint trigger (for
-- the FOREIGN KEY) and then drop the tables, which will also drop the 
-- constraint triggers.
CREATE TABLE baz(a INTEGER PRIMARY KEY, b TEXT, c BOOLEAN);
CREATE TABLE xyzzy(d INTEGER NOT NULL REFERENCES baz);
DROP TABLE baz CASCADE;

-- Clean up.
DROP EVENT TRIGGER trigcreate;
DROP FUNCTION on_trigger_create();
DROP FUNCTION ignore();
DROP EXTENSION schema_triggers;
