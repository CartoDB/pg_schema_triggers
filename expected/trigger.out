CREATE EXTENSION schema_triggers;
-- Create an event trigger for the trigger_create event.
CREATE FUNCTION on_trigger_create()
 RETURNS event_trigger
 LANGUAGE plpgsql
 AS $$
	DECLARE
		event_info SCHEMA_TRIGGERS.TRIGGER_CREATE_EVENTINFO;
	BEGIN
		event_info := schema_triggers.get_trigger_create_eventinfo();
		RAISE NOTICE 'on_trigger_create: (tgname=''%'', tgrelid=''%'', tgenabled=''%'', is_internal=''%'')',
			(event_info.new).tgname, (event_info.new).tgrelid::REGCLASS,
			(event_info.new).tgenabled, event_info.is_internal;
		IF (event_info.new).tgname LIKE 'test_%' THEN
			RAISE EXCEPTION 'trigger name cannot begin with "test_"';
		END IF;
		IF (event_info.new).tgrelid = 'untriggered'::REGCLASS THEN
			RAISE EXCEPTION 'cannot add triggers to relation named "untriggered".';
		END IF;
	END;
$$;
CREATE EVENT TRIGGER trigcreate ON trigger_create
	EXECUTE PROCEDURE on_trigger_create();
-- Create an event trigger for the trigger_adjust event.
-- TODO
-- Create an event trigger for the trigger_rename event.
-- TODO
-- Create an event trigger for the trigger_drop event.
CREATE FUNCTION on_trigger_drop()
 RETURNS event_trigger
 LANGUAGE plpgsql
 AS $$
	DECLARE
		event_info SCHEMA_TRIGGERS.TRIGGER_DROP_EVENTINFO;
	BEGIN
		event_info := schema_triggers.get_trigger_drop_eventinfo();
		IF (event_info.old).tgisinternal IS TRUE THEN
			-- Internal triggers (e.g. for enforcing foreign keys) include an
			-- Oid value in their automatically-generated name;  thus we are
			-- unable to include this in the regression test's output...
			RAISE NOTICE 'on_trigger_drop: dropping internal trigger.';
		ELSE
			RAISE NOTICE 'on_trigger_drop: (tgname=''%'', tgrelid=''%'', tgenabled=''%'')',
				(event_info.old).tgname, (event_info.old).tgrelid::REGCLASS,
				(event_info.old).tgenabled;
		END IF;
	END;
$$;
CREATE EVENT TRIGGER trigdrop ON trigger_drop
	EXECUTE PROCEDURE on_trigger_drop();
-- TODO
-- Create some empty tables.
CREATE TABLE foo();
CREATE TABLE foobar();
CREATE TABLE untriggered();
-- Create a do-nothing trigger function for the following tests.
CREATE FUNCTION ignore()
 RETURNS trigger
 LANGUAGE plpgsql
 AS $$ BEGIN RETURN NULL; END; $$;
-- Create some triggers on the tables.
CREATE TRIGGER foo_trig BEFORE INSERT ON foo
	EXECUTE PROCEDURE ignore();
NOTICE:  on_trigger_create: (tgname='foo_trig', tgrelid='foo', tgenabled='O', is_internal='f')
CREATE TRIGGER foobar_trig BEFORE UPDATE OR DELETE ON foobar
	EXECUTE PROCEDURE ignore();
NOTICE:  on_trigger_create: (tgname='foobar_trig', tgrelid='foobar', tgenabled='O', is_internal='f')
CREATE TRIGGER test_will_not_work BEFORE INSERT ON foobar
	EXECUTE PROCEDURE ignore();
NOTICE:  on_trigger_create: (tgname='test_will_not_work', tgrelid='foobar', tgenabled='O', is_internal='f')
ERROR:  trigger name cannot begin with "test_"
CREATE TRIGGER also_will_not_work BEFORE INSERT ON untriggered
	EXECUTE PROCEDURE ignore();
NOTICE:  on_trigger_create: (tgname='also_will_not_work', tgrelid='untriggered', tgenabled='O', is_internal='f')
ERROR:  cannot add triggers to relation named "untriggered".
-- Rename, enable, and disable the triggers.
ALTER TABLE foo DISABLE TRIGGER foo_trig;
ALTER TRIGGER foo_trig ON foo RENAME TO foo_trig2;
ALTER TABLE foo ENABLE TRIGGER foo_trig2;
ALTER TABLE foobar ENABLE REPLICA TRIGGER foobar_trig;
ALTER TABLE foobar ENABLE ALWAYS TRIGGER foobar_trig;
ALTER TABLE foobar DISABLE TRIGGER foobar_trig;
-- Drop the triggers.
DROP TRIGGER foo_trig2 ON foo;
NOTICE:  on_trigger_drop: (tgname='foo_trig2', tgrelid='foo', tgenabled='O')
DROP TRIGGER foobar_trig ON foobar;
NOTICE:  on_trigger_drop: (tgname='foobar_trig', tgrelid='foobar', tgenabled='D')
-- Create a table with an automatically-created constraint trigger (for
-- the FOREIGN KEY) and then drop the tables, which will also drop the 
-- constraint triggers.
CREATE TABLE baz(a INTEGER PRIMARY KEY, b TEXT, c BOOLEAN);
CREATE TABLE xyzzy(d INTEGER NOT NULL REFERENCES baz);
DROP TABLE baz CASCADE;
NOTICE:  drop cascades to constraint xyzzy_d_fkey on table xyzzy
NOTICE:  on_trigger_drop: dropping internal trigger.
NOTICE:  on_trigger_drop: dropping internal trigger.
NOTICE:  on_trigger_drop: dropping internal trigger.
NOTICE:  on_trigger_drop: dropping internal trigger.
-- Clean up.
DROP EVENT TRIGGER trigcreate;
DROP EVENT TRIGGER trigdrop;
DROP FUNCTION on_trigger_create();
DROP FUNCTION on_trigger_drop();
DROP FUNCTION ignore();
DROP EXTENSION schema_triggers;
